core/src/test/java/org/apache/iceberg/TestMetrics.java

```diff
@@ -347,14 +327,120 @@ public void testMetricsForNullColumns() throws IOException {
+
+    Record nanRecord = GenericRecord.create(FLOAT_DOUBLE_ONLY_SCHEMA);
+    nanRecord.setField("floatCol", Float.NaN);
+    nanRecord.setField("doubleCol", Double.NaN);
+
+    Metrics metrics = getMetrics(FLOAT_DOUBLE_ONLY_SCHEMA, nanRecord, nonNaNRecord1, nonNaNRecord2);
+    Assert.assertEquals(3L, (long) metrics.recordCount());
+    assertCounts(1, 3L, 0L, 1L, metrics);
+    assertCounts(2, 3L, 0L, 1L, metrics);
+
+    // below: current behavior; will be non-NaN values once NaN is excluded from upper/lower bound. ORC and Parquet's
+    // behaviors differ due to their implementation of comparison being different.
+    if (fileFormat() == FileFormat.ORC) {
+      assertBounds(1, FloatType.get(), Float.NaN, Float.NaN, metrics);
+      assertBounds(2, DoubleType.get(), Double.NaN, Double.NaN, metrics);
+    } else {
+      assertBounds(1, FloatType.get(), 1.2F, Float.NaN, metrics);
+      assertBounds(2, DoubleType.get(), 3.4D, Double.NaN, metrics);
+    }
+  }
+
+  @Test
+  public void testColumnBoundsWithNaNValueInMiddle() throws IOException {
+    Record nonNaNRecord1 = GenericRecord.create(FLOAT_DOUBLE_ONLY_SCHEMA);
+    nonNaNRecord1.setField("floatCol", 1.2F);
+    nonNaNRecord1.setField("doubleCol", 3.4D);
+
+    Record nonNaNRecord2 = GenericRecord.create(FLOAT_DOUBLE_ONLY_SCHEMA);
+    nonNaNRecord2.setField("floatCol", 5.6F);
+    nonNaNRecord2.setField("doubleCol", 7.8D);
+
+    Record nanRecord = GenericRecord.create(FLOAT_DOUBLE_ONLY_SCHEMA);
+    nanRecord.setField("floatCol", Float.NaN);
+    nanRecord.setField("doubleCol", Double.NaN);
+
+    Metrics metrics = getMetrics(FLOAT_DOUBLE_ONLY_SCHEMA, nonNaNRecord1, nanRecord, nonNaNRecord2);
+    Assert.assertEquals(3L, (long) metrics.recordCount());
+    assertCounts(1, 3L, 0L, 1L, metrics);
+    assertCounts(2, 3L, 0L, 1L, metrics);
+
+    // below: current behavior; will be non-NaN values once NaN is excluded from upper/lower bound. ORC and Parquet's
+    // behaviors differ due to their implementation of comparison being different.
+    if (fileFormat() == FileFormat.ORC) {
+      assertBounds(1, FloatType.get(), 1.2F, 5.6F, metrics);
+      assertBounds(2, DoubleType.get(), 3.4D, 7.8D, metrics);
+    } else {
+      assertBounds(1, FloatType.get(), 1.2F, Float.NaN, metrics);
+      assertBounds(2, DoubleType.get(), 3.4D, Double.NaN, metrics);
+    }
+  }
+
+  @Test
+  public void testColumnBoundsWithNaNValueAtEnd() throws IOException {
+    Record nonNaNRecord1 = GenericRecord.create(FLOAT_DOUBLE_ONLY_SCHEMA);
+    nonNaNRecord1.setField("floatCol", 1.2F);
+    nonNaNRecord1.setField("doubleCol", 3.4D);
+
+    Record nonNaNRecord2 = GenericRecord.create(FLOAT_DOUBLE_ONLY_SCHEMA);
+    nonNaNRecord2.setField("floatCol", 5.6F);
+    nonNaNRecord2.setField("doubleCol", 7.8D);
+
+    Record nanRecord = GenericRecord.create(FLOAT_DOUBLE_ONLY_SCHEMA);
+    nanRecord.setField("floatCol", Float.NaN);
+    nanRecord.setField("doubleCol", Double.NaN);
```

Comment:
You can also make these records static variables. That might reduce the size of these test cases.